exec/cnex/cnex.c

```diff
@@ -28,28 +30,88 @@ typedef struct tagTCommandLineOptions {
     int  ArgStart;
     char *pszFilename;
     char *pszExecutableName;
+    char *pszExecutablePath;
 } TOptions;
 
-static TOptions gOptions = { FALSE, FALSE, TRUE, 0, NULL, NULL };
+static TOptions gOptions = { FALSE, FALSE, TRUE, 0, NULL, NULL, NULL };
 
-void exec_freeExecutor(TExecutor *e)
+static char *getApplicationName(char *arg)
 {
-    uint32_t i;
+    char *p = arg;
+    char *s = p;
+
+    while (*p) {
+        if (*p == '\\' || *p == '/') {
+            s = p;
+        }
+        p++;
+    }
+    s++;
+    return s;
+}
 
-    for (i = 0; i < e->object->global_size; i++) {
-        cell_clearCell(&e->globals[i]);
+static char *getApplicationPath(char *arg)
+{
+    BOOL bHavePath = TRUE;
+    char *p = arg;  // p pointer to provided arg.
+    char *e = p;    // e Last path seperator found.
+
+    // Walk the arg string, recording the last position of a path seperator char.
+    while (*p) {
+        if (*p == '\\' || *p == '/' || *p == ':') {
+            e = p;
+        }
+        p++;
     }
-    free(e->globals);
+
+    if (e == arg) {
+        // We didn't find any path seperators.  So Set the return path to simply be the "." current dir.
+        bHavePath = FALSE;
+    }
+
+    // Allocate space for our return value.  Note: We add two extra bytes here for the path.  Since we walked
+    // to the last path seperator, we need one byte for it, plus a byte for our NULL char.
+    char *r = malloc((e - arg) + 2);
```

Comment:
If you're going to increment `e` anyway in each branch of the following if statement, why not increment it up here, before the malloc, so you don't need to explain why you're adding 2 and not 1 byte.

Reply:
Fixed.